multi: Rework utxoset/view to use outpoints.

rstaudt2 opened the pull request:
**NOTE: This contains a database migration, so if you plan to test it, please make a copy of your data directory first.**

---

This modifies the utxoset in the database and related `UtxoViewpoint` to store and work with unspent transaction outputs on a per-output basis instead of at a transaction level.  This is a port of https://github.com/btcsuite/btcd/pull/1045 with some significant differences due to the stake tree and transactions in Decred.

The primary motivation is to simplify the code, pave the way for a utxo cache, and generally focus on optimizing runtime performance.

The tradeoff is that this approach does somewhat increase the size of the serialized utxoset since it means that the transaction hash is duplicated for each output as a part of the key and some additional details are duplicated in each output.  The details duplicated in each output include flags encoded into a single byte that specify whether the containing transaction is a coinbase, whether the containing transaction has an expiry, and the transaction type.  Additionally, the containing block height and index are stored in each output.

However, in practice, the size difference isn't all that large, disk space is relatively cheap, certainly cheaper than memory, and it is much more important to provide more efficient runtime operation since that is the ultimate purpose of the daemon.

While performing this conversion, it also simplifies the code to remove the transaction version information from the utxoset as well as the spend journal.  The logic for only serializing it under certain circumstances is complicated, and it was only used for the `gettxout` RPC, where it has already been removed.

The utxo set and spend journal in the database are automatically migrated to the new format with this commit and it is possible to interrupt and resume the migration process.

Finally, it also updates all references and tests that previously dealt with transaction hashes to use outpoints instead.

An overview of the changes are as follows:

- Remove transaction version from both spent and unspent output entries
  - Update utxo serialization format to exclude the version
  - Update spend journal serialization format to exclude the version
- Convert `UtxoEntry` to represent a specific utxo instead of a transaction with all remaining utxos
  - Optimize for memory usage with an eye toward a utxo cache
    - Combine fields such as whether the containing transaction is a coinbase, whether the containing transaction has an expiry, and the transaction type, into a single byte
    - Align entry fields to eliminate extra padding since ultimately there will be a lot of these in memory
    - Introduce a free list for serializing an outpoint to the database key format to significantly reduce pressure on the GC
  - Update entries to be keyed by a `<hash><tree><output index>` outpoint rather than just a tx hash
  - Update all related functions that previously dealt with transaction hashes to accept outpoints instead
  - Update all callers accordingly
  - Only add individually requested outputs from the mempool when constructing a mempool view
- Modify the spend journal to always store the encoded flags with every spent txout
  - Combine fields such as whether the containing transaction is a coinbase, whether the containing transaction has an expiry, and the transaction type into a single byte
    - Use 4 bits instead of 3 for the transaction type to be consistent with utxos. The extra bit was already unused so this doesn't take any additional space
  - Remove the fully spent flag
- Introduce `ticketMinOuts` in place of `stakeExtra`
  - Renamed `stakeExtra` as `ticketMinOuts` and updated all comments to make the purpose of the field clearer
  - Only store `ticketMinOuts` for ticket submission outputs
  - Add `TicketMinimalOutputs` function on `UtxoEntry` in place of `ConvertUtxosToMinimalOutputs`
- Always decompress data loaded from the database now that a utxo entry only consists of a specific output
- Introduce upgrade code to migrate the utxo set and spend journal to the new format
  - Update current database version to 9
  - Update current utxo set version to 3
  - Update current spend journal version to 3
  - Introduce the ability to run upgrades after the block index has been loaded
- Update all tests to expect the correct encodings, remove tests that no longer apply, and add new ones for the new expected behavior
  - Convert old tests for the legacy utxo format deserialization code to test the new function that is used during upgrade
- Introduce a few new functions on `UtxoViewpoint`
  - `AddTxOut` for adding an individual txout versus all of them
  - `addTxOut` to handle the common code between the new `AddTxOut` and existing `AddTxOuts`
  - `RemoveEntry` for removing an individual txout
- Remove the `ErrDiscordantTxTree` error
  - Since utxos are now retrieved using an outpoint, which includes the tree, it is no longer possible to hit this error path

---

A couple other notes on the implementation:

- Unfortunately, it wasn't really possible to break this into smaller working commits, so these changes are all in one large commit.
- The spend journal migration does take awhile to run (~20 mins, due to the fact that it needs to read all block entries).

---

**Part of https://github.com/decred/dcrd/issues/1145.**

davecgh reviewed (changes requested):
Round 1.

I still have a lot of review and testing to do, but I gave this a very quick first pass to find anything that stood out before really digging into the details and doing in-depth testing that I wanted to go ahead and provide feedback for first.

This is fantastic overall and great to see it ported over.  I really appreciate the time put into creating the PR description with all of the relevant details and references, important notes, testing implications, etc.  That really helps not only the initial review, but for historical purposes as well.  Also, from having done the PR this one is porting over, I know how much effort it takes to make this type of changes that have far reaching implications and thus also make a big impact, so I thank you for taking this on!

rstaudt2 commented on blockchain/fullblocktests/generate.go:
```diff
@@ -2003,19 +2003,6 @@ func Generate(includeLargeReorg bool) (tests [][]TestInstance, err error) {
-		tx.TxIn[0].PreviousOutPoint.Tree = wire.TxTreeStake
-		b.AddTransaction(tx)
-	})
-	rejected(blockchain.ErrDiscordantTxTree)
```
Good point, I added the test back with the expected `ErrMissingTxOut` error.

rstaudt2 commented on blockchain/scriptval.go:
```diff
@@ -60,7 +60,7 @@ out:
 			originTxHash := &txIn.PreviousOutPoint.Hash
 			originTxIndex := txIn.PreviousOutPoint.Index
-			txEntry := v.utxoView.LookupEntry(originTxHash)
+			txEntry := v.utxoView.LookupEntry(txIn.PreviousOutPoint)
```
Updated these error cases.

rstaudt2 commented on blockchain/go.mod:
```diff
@@ -3,6 +3,7 @@ module github.com/decred/dcrd/blockchain/v4
 go 1.13
 
 require (
+	github.com/davecgh/go-spew v1.1.1
```
Removed.

rstaudt2 commented on internal/mempool/mempool_test.go:
```diff
@@ -83,16 +83,26 @@ func (s *fakeChain) FetchUtxoView(tx *dcrutil.Tx, treeValid bool) (*blockchain.U
-	// Add an entry for the tx itself to the new view.
+	msgTx := tx.MsgTx()
+	txType := stake.DetermineTxType(msgTx, noTreasury)
+	tree := wire.TxTreeRegular
```
Updated to use `tx.Tree()`.

rstaudt2 commented on internal/mining/mining_harness_test.go:
```diff
@@ -140,16 +140,26 @@ func (c *fakeChain) FetchUtxoView(tx *dcrutil.Tx, treeValid bool) (*blockchain.U
-	// Add an entry for the tx itself to the new view.
+	msgTx := tx.MsgTx()
+	txType := stake.DetermineTxType(msgTx, c.isTreasuryAgendaActive)
+	tree := wire.TxTreeRegular
```
Updated to use `tx.Tree()`.

davecgh replied:
Did this get reverted?  I see it commented that it was updated, but it doesn't appear to be.

rstaudt2 replied:
I thought I had updated this both here and in the mempool test harness, but it looks like it was missed here.  Updated now.

rstaudt2 commented on internal/mining/mining.go:
```diff
@@ -749,16 +747,14 @@ func (g *BlkTmplGenerator) maybeInsertStakeTx(stx *dcrutil.Tx, treeValid bool, i
-		originHash := &txIn.PreviousOutPoint.Hash
-		utxIn := view.LookupEntry(originHash)
-		if utxIn == nil {
+		utxoIn := view.LookupEntry(txIn.PreviousOutPoint)
```
Good call, I updated the naming here.

rstaudt2 commented on internal/mining/mining.go:
```diff
@@ -1970,17 +1965,15 @@ nextPriorityQueueItem:
-			originHash := &txIn.PreviousOutPoint.Hash
-			utx := utxs.LookupEntry(originHash)
-			if utx == nil {
+			utxo := view.LookupEntry(txIn.PreviousOutPoint)
```
Updated.

rstaudt2 commented on internal/rpcserver/rpcserver.go:
```diff
@@ -3539,6 +3543,11 @@ func handleGetTxOut(_ context.Context, s *Server, cmd interface{}) (interface{},
 	}
 	if txFromMempool != nil {
+		// Return nil if the tx tree does not match the tree param that was passed.
+		if txFromMempool.Tree() != c.Tree {
```
Good call, should really never happen in practice but agreed.  Updated.

rstaudt2 commented on blockchain/validate.go:
```diff
@@ -1819,7 +1819,11 @@ func (b *BlockChain) checkDupTxs(txSet []*dcrutil.Tx, view *UtxoViewpoint) error
 	// Typically, there will not be any utxos for any of the transactions.
```
Updated.

rstaudt2 commented on internal/rpcserver/rpcserverhandlers_test.go:
```diff
@@ -56,67 +56,68 @@ const testDataPath = "testdata"
 	isCoinBase           bool
-	isOutputSpent        bool
+	isSpent              bool
 	modified             bool
```
It's actually used in `gettxout` here (both before and after this PR): https://github.com/decred/dcrd/blob/26cdfe6aae1c4ecbcf0a49809925a1caf11f1c77/internal/rpcserver/rpcserver.go#L3583-L3591.  So I think it probably is needed if we want to be able to mock and test that code path.

davecgh replied:
I was commenting on the `modified` field, not the updated `isSpent` field.  I should've been more explicit in the comment since it's right next to the one that was actually changed, I can see how that's hard to notice.

rstaudt2 replied:
Ah, ya, I definitely saw the diff and thought you were talking about the `isSpent` field.  Makes sense for the `modified` field.  It was actually only referenced in one place, which wasn't even needed, so I removed it.

rstaudt2 commented on blockchain/validate.go:
```diff
@@ -2446,8 +2452,8 @@ func checkRevocationInputs(tx *dcrutil.Tx, txHeight int64, view *UtxoViewpoint,
-	ticketOuts := ConvertUtxosToMinimalOutputs(ticketUtxo)
-	if len(ticketOuts) == 0 {
+	ticketOuts := ticketUtxo.TicketMinimalOutputs()
+	if ticketOuts == nil {
```
Updated, though I don't think the case of it having a zero length and not being nil is possible, since if there were no outputs then `ticketUtxo` itself would be nil.

davecgh replied:
Yeah, I agree with the current code that is true, but this is an assertion to try and catch bugs as things evolve.  I could see a case later were a utxo entry is reused in the cache and the assumption gets violated where it forgets to nil it out and instead ends up with a slice that has been cleared.

rstaudt2 commented on blockchain/validate.go:
```diff
@@ -2681,28 +2688,12 @@ func CheckTransactionInputs(subsidyCache *standalone.SubsidyCache, tx *dcrutil.T
 			scriptClass := txscript.GetScriptClass(
-				utxoEntry.ScriptVersionByIndex(originTxIndex),
-				utxoEntry.PkScriptByIndex(originTxIndex),
+				utxoEntry.ScriptVersion(),
```
Good call, I updated this and the below cases.

rstaudt2 commented on blockchain/utxoviewpoint.go:
```diff
@@ -17,184 +17,169 @@ import (
 
-// BlockIndex returns the height of the block containing the transaction the
-// utxo entry represents.
+// BlockIndex returns the index of the block containing the output.
```
Whoops, ya, that was worded poorly and has been updated.

rstaudt2 commented on blockchain/utxoviewpoint.go:
```diff
@@ -265,31 +235,136 @@ func (view *UtxoViewpoint) SetBestHash(hash *chainhash.Hash) {
+	entry.blockIndex = blockIndex
+	entry.scriptVersion = txOut.Version
+	entry.packedFlags = packedFlags
+	if isTicketSubmissionOutput(entry.TransactionType(), outpoint.Index) {
```
Agreed, removed.

rstaudt2 commented on blockchain/utxoviewpoint.go:
```diff
@@ -265,31 +235,136 @@ func (view *UtxoViewpoint) SetBestHash(hash *chainhash.Hash) {
-	return entry
+	// Set encoded flags for the transaction.
+	isCoinBase := standalone.IsCoinBaseTx(msgTx, isTreasuryEnabled)
+	spent := false
```
Updated this case (and others) to constants.

rstaudt2 commented on blockchain/utxoviewpoint.go:
```diff
@@ -265,31 +235,136 @@ func (view *UtxoViewpoint) SetBestHash(hash *chainhash.Hash) {
+
+	// Set encoded flags for the transaction.
+	isCoinBase := standalone.IsCoinBaseTx(msgTx, isTreasuryEnabled)
+	spent := false
```
Updated.

rstaudt2 commented on blockchain/utxoviewpoint.go:
```diff
@@ -298,81 +373,22 @@ func (view *UtxoViewpoint) PrevScript(prevOut *wire.OutPoint) (uint16, []byte, b
-	txEntry := view.LookupEntry(originHash)
-	if txEntry != nil && !txEntry.IsOutputSpent(originIndex) {
-		return txEntry.BlockHeight(), txEntry.AmountByIndex(originIndex), true
+	txEntry := view.LookupEntry(*prevOut)
```
Renamed this and a few other cases.

rstaudt2 commented on blockchain/utxoviewpoint.go:
```diff
@@ -405,47 +420,39 @@ func (view *UtxoViewpoint) connectTransaction(tx *dcrutil.Tx, blockHeight int64,
+		// Only create the stxo details if requested.
+		if stxos != nil {
+			// Populate the stxo details using the utxo entry.
+			spent := true
```
Updated.

rstaudt2 commented on blockchain/utxoviewpoint.go:
```diff
@@ -405,47 +420,39 @@ func (view *UtxoViewpoint) connectTransaction(tx *dcrutil.Tx, blockHeight int64,
-
-			if entry.txType == stake.TxTypeSStx {
-				stxo.stakeExtra = entry.stakeExtra
+			if entry.ticketMinOuts != nil {
```
Agreed, combined it as suggested.

rstaudt2 commented on blockchain/utxoviewpoint.go:
```diff
@@ -488,25 +549,61 @@ func (view *UtxoViewpoint) disconnectTransactions(block *dcrutil.Block, stxos []
+			outpoint.Index = uint32(txOutIdx)
+			entry := view.entries[outpoint]
+			if entry == nil {
+				spent := false
```
Updated.

rstaudt2 commented on blockchain/utxoviewpoint.go:
```diff
@@ -515,72 +612,93 @@ func (view *UtxoViewpoint) disconnectTransactions(block *dcrutil.Block, stxos []
 				}
-				view.entries[*originHash] = entry
+
+				spent := true
```
Updated.

rstaudt2 commented on blockchain/utxoviewpoint.go:
```diff
@@ -749,65 +866,63 @@ func (view *UtxoViewpoint) disconnectBlock(db database.DB, block, parent *dcruti
-	// since other code uses the presence of an entry in the store as a way
-	// to optimize spend and unspend updates to apply only to the specific
-	// utxos that the caller needs access to.
+	// so other code can use the presence of an entry in the store as a way
```
Fixed.

rstaudt2 commented on blockchain/upgrade.go:
```diff
@@ -1789,6 +1789,384 @@ func upgradeToVersion8(ctx context.Context, db database.DB, dbInfo *databaseInfo
+
+	// Migrate the utxoset to version 3.
+	v3DoneKeyName := []byte("utxosetv3done")
+	err := runUpgradeStageOnce(ctx, db, v3DoneKeyName, func() error {
```
Updated to not run as a stage.

rstaudt2 commented on blockchain/upgrade.go:
```diff
@@ -1789,6 +1789,384 @@ func upgradeToVersion8(ctx context.Context, db database.DB, dbInfo *databaseInfo
+	// Update and persist the database versions and remove upgrade progress
+	// tracking keys.
+	err = db.Update(func(dbTx database.Tx) error {
+		err := dbTx.Metadata().Delete(v3DoneKeyName)
```
Removed.

rstaudt2 commented on blockchain/upgrade.go:
```diff
@@ -1789,6 +1789,384 @@ func upgradeToVersion8(ctx context.Context, db database.DB, dbInfo *databaseInfo
+		return err
+	}
+
+	if interruptRequested(ctx) {
```
Removed.

rstaudt2 commented on blockchain/upgrade.go:
```diff
@@ -1789,6 +1789,384 @@ func upgradeToVersion8(ctx context.Context, db database.DB, dbInfo *databaseInfo
+
+			// Add an entry for each utxo into the new bucket using the new format.
+			for txOutIdx, utxo := range utxos {
+				reserialized, err := serializeUtxoEntry(utxo)
```
I agree with this and will update it.

rstaudt2 replied:
This has been updated to just deal with the raw data and not rely on any of the serialize/deserialize functions.

rstaudt2 commented on internal/mempool/mempool.go:
```diff
@@ -1116,18 +1116,24 @@ func (mp *TxPool) fetchInputUtxos(tx *dcrutil.Tx, isTreasuryEnabled bool) (*bloc
-	for originHash, entry := range utxoView.Entries() {
-		if entry != nil && !entry.IsFullySpent() {
+	for _, txIn := range tx.MsgTx().TxIn {
+		prevOut := &txIn.PreviousOutPoint
```
I renamed `prevOut` to `outpoint` in the cases where the variable referred to an output of a transaction, as opposed to a previous outpoint associated with a transaction input (I found the naming `prevOut` to be misleading in that case).  The cases where it was actually referring to a previous outpoint I left as is.  Given that, do you still think it would be best to rename all of them to `outpoint`?  I don't have a strong opinion and would be good with it either way.

davecgh replied:
Makes sense.  I similarly don't feel strongly either way myself.  I suppose it really just depends on how it's going to be used as outpoints ultimately become prev outs, so if this make it more clear to you, works for me!

rstaudt2 replied:
Sounds good.  In that case, I'll leave it as is since I think it's _slightly_ more clear, especially in the cases where a function is dealing with both the inputs and outputs of a given transaction.

rstaudt2 commented on blockchain/utxoviewpoint.go:
```diff
@@ -454,15 +461,68 @@ func (view *UtxoViewpoint) connectTransaction(tx *dcrutil.Tx, blockHeight int64,
+	trees := []int8{wire.TxTreeRegular, wire.TxTreeStake}
+	for _, tree := range trees {
+		outpoint := wire.OutPoint{Hash: *hash, Tree: tree}
+		for idx := uint32(0); idx < 3; idx++ {
```
Good catch in regards to the entry loaded from the database being unmodified.  I Will update this to check all possible outputs here.

rstaudt2 replied:
This has been updated to use a pre-calculated `maxLegacyOutputsPerBlock` const to check all possible outputs.

As already mentioned in the function and inline comments, this is super inefficient but should really never happen.  Also, in the case it should happen, it will return immediately after finding the output in the view (which will be an index much lower than `maxLegacyOutputsPerBlock` and would only go to `maxLegacyOutputsPerBlock` in the worst possible case or if the output does not exist in the view).

As already mentioned in the PR description, we would need to migrate the data to avoid this, but there are some complications with attempting to do that efficiently.  One alternative to migrating would be to iterate all blocks and create the spend journal entries from scratch in the new format.  However, this requires context checks, such as whether the treasury agenda is active to determine the tx type, so it would be slightly different from current migrations and would have to run after the block index is loaded. I think this will probably end up being the best approach, and I can follow up on it separately from this PR.

rstaudt2 commented on blockchain/utxoviewpoint.go:
```diff
@@ -17,184 +17,169 @@ import (
+	//
+	// This approach is used since all outputs of a ticket need to be retrieved
+	// when validating vote transaction inputs or revoking a ticket.
+	ticketMinOuts []byte
```
We discussed this a bit separately but adding some detail here for reference.

I considered an alternative approach to this where we would store all ticket outputs in the view (as separate utxo entries) and then retrieve them all as needed (when validating vote transaction inputs or revoking a ticket).  This would minimize the duplication of data being stored, but it led to some complexity for a few reasons:

- You need to add all ticket outputs to the view any time you add a ticket submission output to the view (which also requires knowing the number of outputs for the ticket, which would need to be stored for both unspent and spent outputs).
- You need to store ticket commitments as unspent and spent outputs (and even serialize them to spend journal entries as spent), which doesn't really make sense since they aren't truly spendable.

To avoid this taking 24 bytes when its `nil` (which will be the case for all outputs other than ticket submissions), I updated this to be a pointer to a struct that holds the data (and therefore will reduce the usage to 8 bytes for a `nil` pointer).  I went with a struct here rather than a pointer to a slice directly since we will likely want to expand what is stored along with that data when implementing the utxo cache.
```
func main() {
	type ticketMinimalOutputs struct {
		data []byte
	}
	type UtxoEntry struct {
		ticketMinOuts *ticketMinimalOutputs
	}
	entry := UtxoEntry{ticketMinOuts: nil}
	fmt.Println(unsafe.Sizeof(entry))
}
```
```
$ go run main.go
8
```

This still leaves plenty of room for improvement, and I'll consider some alternative approaches as part of the utxo cache work.  I think ultimately, we will want separate views for the regular and stake trees, which we have discussed previously and should allow us to simplify some scenarios like that.

rstaudt2 commented on blockchain/utxoviewpoint.go:
```diff
@@ -515,72 +612,93 @@ func (view *UtxoViewpoint) disconnectTransactions(block *dcrutil.Block, stxos []
-				if !stxo.txFullySpent {
-					return AssertError(fmt.Sprintf("tried to revive unspent "+
-						"tx %v from non-fully spent stx entry", originHash))
+			if !stxo.IsSpent() {
```
I know you are already familiar with this but just adding some context here.  The only case where `stxo.IsSpent()` would return `false` is for legacy v2 spend journal entries that were **not** the last spent output of the containing transaction (since the spent flag used to represent _fully_ spent).  This case is the only scenario where the `stxo` does not have the encoded flags stored, and therefore the flags need to be retrieved from a utxo for that transaction.

rstaudt2 commented on blockchain/validate.go:
```diff
@@ -1829,12 +1833,16 @@ func (b *BlockChain) checkDupTxs(txSet []*dcrutil.Tx, view *UtxoViewpoint) error
-				tx.Hash(), txEntry.BlockHeight())
-			return ruleError(ErrOverwriteTx, str)
+		outpoint := wire.OutPoint{Hash: *tx.Hash(), Tree: tree}
+		for txOutIdx := range tx.MsgTx().TxOut {
```
Ya, I agree, that is a good point.  I'll take a look at this in a separate PR.

rstaudt2 commented on blockchain/utxoviewpoint.go:
```diff
@@ -265,31 +235,136 @@ func (view *UtxoViewpoint) SetBestHash(hash *chainhash.Hash) {
+	}
+
+	entry.amount = txOut.Value
+	entry.pkScript = txOut.PkScript
```
Noted, I'll consider this with the utxo cache implementation.

rstaudt2 commented:
I made an update to store `ticketMinOuts` in all ticket outputs rather than only in ticket submission outputs.  This was required because the legacy spend journal stored that data for every ticket output, so to maintain backward compatibility, it needs to remain the same (otherwise, there will be an issue in the unlikely case of deep reorgs). 

I am going to start taking a look at implementing a migration that recreates the spend journal data in the new format from scratch (in a separate PR) since I'd prefer not to have to keep the legacy handling code around for long.

dnldd commented on internal/mining/mining_harness_test.go:
```diff
@@ -678,7 +682,7 @@ func (p *fakeTxSource) maybeAcceptTransaction(tx *dcrutil.Tx, isNew bool) ([]*ch
 			}
-			if entry.IsFullySpent() {
+			if entry.IsSpent() {
 				log.Tracef("Transaction %v uses full spent input %v "+
```
s / ... uses full spent input ... / ... uses spent input ...

rstaudt2 replied:
Good call, fixed this case and a similar case in `mempool_test.go`.

dnldd commented on internal/netsync/manager.go:
```diff
@@ -1793,41 +1821,80 @@ func (m *SyncManager) requestFromPeer(p *peerpkg.Peer, blocks, txs []*chainhash.
+			}
+
+			// Check if the transaction exists from the point of view of the end of
+			// the main chain.  Note that this is only a best effort since it is
```
s /  ... from the point of view of the end of the main chain. / ... from the point of view of the main chain tip.

rstaudt2 replied:
Updated.

dnldd commented on rpcadaptors.go:
```diff
@@ -395,25 +394,18 @@ type rpcChain struct {
-// FetchUtxoEntry loads and returns the unspent transaction output entry for the
-// passed hash from the point of view of the end of the main chain.
+// FetchUtxoEntry loads and returns the requested unspent transaction output
+// from the point of view of the end of the main chain.
```
s / ... from the point of view of the end of the main chain / ... from the point of view of the main chain tip.

rstaudt2 replied:
Changed.

rstaudt2 commented:
Found one more issue related to processing deep reorgs with the legacy stxo handling ([diff](https://github.com/decred/dcrd/compare/87a1d67131e1ab09b95a5752a2fa11380d22fba6..335e895109a67d198aaf1973f982ce77ebbc61ef)).  This is actually a bug in the [btcd code](https://github.com/btcsuite/btcd/blob/6bd4c64a54faeb343d5e3c1ee5802450a291787f/blockchain/utxoviewpoint.go#L425) as well - it is fetching a utxo using the tx hash rather than the hash of the input transaction (though it doesn't really matter, since that code path would never be hit at this point since its only run when encountering legacy stxo entires).

I should be able to get the stxo migration in place relatively quickly, which will avoid this case altogether (though unfortunately, since it needs to read all historical blocks/txns, it will likely take in the range of 10-20 mins to complete).

rstaudt2 commented on blockchain/upgrade.go:
```diff
@@ -1789,6 +1789,398 @@ func upgradeToVersion8(ctx context.Context, db database.DB, dbInfo *databaseInfo
+					encodedFlags |= 1
+				}
+				if hasExpiry {
+					encodedFlags |= 1 << 2
```
Good catch!  Fixed.

rstaudt2 commented on blockchain/upgrade.go:
```diff
@@ -1789,6 +1789,398 @@ func upgradeToVersion8(ctx context.Context, db database.DB, dbInfo *databaseInfo
+				copy(ticketMinOuts, oldSerialized[offset:])
+			}
+
+			// Create a V3 utxo entries with the details deserialized above.
```
Good call, I added a section above to outline the V3 format similar to the V2 version.

rstaudt2 commented on blockchain/upgrade.go:
```diff
@@ -1789,6 +1789,398 @@ func upgradeToVersion8(ctx context.Context, db database.DB, dbInfo *databaseInfo
+				//   bit  0     - containing transaction is a coinbase
+				//   bit  1     - containing transaction has an expiry
+				//   bits 2-5   - transaction type
+				//   bit  6     - output is spent
```
As part of adding more detail for the V3 format above, I updated this to say `unused` instead since the bit will never be set (and nothing is requiring the encoding of the flags to be the same between utxos/stxos other than it being convenient for encoding/decoding).

rstaudt2 commented on blockchain/upgrade.go:
```diff
@@ -1789,6 +1789,398 @@ func upgradeToVersion8(ctx context.Context, db database.DB, dbInfo *databaseInfo
+				// Encode the V3 utxo flags.  The bit representation is:
+				//   bit  0     - containing transaction is a coinbase
+				//   bit  1     - containing transaction has an expiry
+				//   bits 2-5   - transaction type
```
That is a fair point, though this is now reserving 4 bits for the tx types, so we won't run out of room until we need more than 16 tx types.  I don't think we will be approaching that any time soon, so I'm not sure that it is worth the effort of moving (not that it would be difficult, especially after the stxo migration is ready, as you mentioned).

sefbkn commented on blockchain/chainio.go:
```diff
@@ -1061,133 +1014,107 @@ func deserializeUtxoEntry(serialized []byte) (*UtxoEntry, error) {
+// dbFetchUtxoEntryByHash attempts to find and fetch a utxo for the given hash.
+// It uses a cursor and seek to try and do this as efficiently as possible.
+//
+// When there are no entries for the provided hash, nil will be returned for the
```
```suggestion
// When there are no entries for the provided hash, nil will be returned for
```

rstaudt2 replied:
Fixed.

sefbkn commented on blockchain/utxoviewpoint.go:
```diff
@@ -265,31 +247,139 @@ func (view *UtxoViewpoint) SetBestHash(hash *chainhash.Hash) {
 // PrevScript returns the script and script version associated with the provided
 // previous outpoint along with a bool that indicates whether or not the
 // requested entry exists.  This ensures the caller is able to distinguish
 // between missing entry and empty v0 scripts.
```
```suggestion
// between missing entries and empty v0 scripts.
```

rstaudt2 replied:
Updated.

sefbkn commented on blockchain/utxoviewpoint.go:
```diff
@@ -298,81 +388,22 @@ func (view *UtxoViewpoint) PrevScript(prevOut *wire.OutPoint) (uint16, []byte, b
-}
-
 // connectTransaction updates the view by adding all new utxos created by the
 // passed transaction and marking all utxos that the transactions spend as
```
```suggestion
// passed transaction and marking all utxos that the transaction spends as
```

rstaudt2 replied:
Fixed.

sefbkn commented on blockchain/utxoviewpoint.go:
```diff
@@ -454,15 +473,58 @@ func (view *UtxoViewpoint) connectTransaction(tx *dcrutil.Tx, blockHeight int64,
+	}
+
+	// Check the database since it doesn't exist in the view.  This will
+	// often by the case since only specifically referenced utxos are loaded
```
```suggestion
	// often be the case since only specifically referenced utxos are loaded
```

rstaudt2 replied:
Updated.

sefbkn commented on blockchain/utxoviewpoint.go:
```diff
@@ -488,25 +551,63 @@ func (view *UtxoViewpoint) disconnectTransactions(block *dcrutil.Block, stxos []
+
+		// Mark all of the spendable outputs originally created by the transaction
+		// as spent.  It is instructive to note that while the outputs aren't
+		// actually being spent here, rather they no longer exist, since a pruned
```
```suggestion
		// actually being spent here, rather they no longer exist.  Since a pruned
```

rstaudt2 replied:
Changed.

sefbkn reviewed (changes requested):
Great work on this PR. Only have a few minor points right now.

rstaudt2 commented:
Rebased to `master` and also made the comment updates that @sefbkn pointed out.

Also, as a heads up, I have the stxo migration ready and will be pushing it to this PR shortly.  Sorry in advance for the churn, but I think it is worth bringing into this PR directly because it avoids having another utxo migration and completely removes the legacy handling cases.  I'll call out the changes/diff when I push that up.

davecgh commented:
> Also, as a heads up, I have the stxo migration ready and will be pushing it to this PR shortly. Sorry in advance for the churn, but I think it is worth bringing into this PR directly because it avoids having another utxo migration and completely removes the legacy handling cases. I'll call out the changes/diff when I push that up.

Agreed.  I was going to suggest this myself if it was done prior to this one being finalized, so glad to see you're of the same mindset.

rstaudt2 commented:
The spend journal migration has now been added ([diff](https://github.com/decred/dcrd/compare/d83712db11c84cf17dd605e079cd4498e4e11777..d138150ee6045867903221189b85946e58925b89)).  The updates include the following:

- Added a `upgradeDBPostBlockIndexLoad` function to run migrations after the block index has been loaded.
- Added the `migrateSpendJournalVersion2To3` migration.  It takes about ~20 mins to run since it has to read all block entries.
- Updated ticket minimal outputs to only be stored in ticket _submission_ outputs (since that could now be handled properly with the spend journal migration in place).
- Removed the reserved field from the spend journal entries since it is no longer needed for backward compatibility.
- Removed the spent flag from the spend journal entries since it is no longer needed for backward compatibility.
- Removed the code that was handling legacy v2 spend journal entries since they are now migrated.

rstaudt2 commented:
> The spend journal migration has now been added ([diff](https://github.com/decred/dcrd/compare/d83712db11c84cf17dd605e079cd4498e4e11777..d138150ee6045867903221189b85946e58925b89)). The updates include the following:
> 
>     * Added a `upgradeDBPostBlockIndexLoad` function to run migrations after the block index has been loaded.
> 
>     * Added the `migrateSpendJournalVersion2To3` migration.  It takes about ~20 mins to run since it has to read all block entries.
> 
>     * Updated ticket minimal outputs to only be stored in ticket _submission_ outputs (since that could now be handled properly with the spend journal migration in place).
> 
>     * Removed the reserved field from the spend journal entries since it is no longer needed for backward compatibility.
> 
>     * Removed the spent flag from the spend journal entries since it is no longer needed for backward compatibility.
> 
>     * Removed the code that was handling legacy v2 spend journal entries since they are now migrated.

I also tested a deep reorg (invalidated/reconsidered a block thousands of blocks deep into the migrated data) to ensure that the migrated spent journal entries could be resurrected properly.

davecgh commented on blockchain/chainio.go:
```diff
@@ -1959,7 +1812,12 @@ func (b *BlockChain) initChainState(ctx context.Context) error {
+	}
+
+	// Upgrade the database post block index load as needed.
+	return upgradeDBPostBlockIndexLoad(ctx, b)
```
We discussed this a bit offline, but I'll comment here for future reference as well.

This will end up causing issues with future multi-version upgrades because it means that the upgrade process will not be able to complete uninterrupted and the need for it to run the normal code which relies on upgrades having happened.

For example, consider an upgrade from say version 7 to a future version 10.  It will complete the version 7 to 8 upgrade, but then not be able to get to version 9 because that can only happen once the normal upgrade process completes and the block index is fully loaded.  However, the code will otherwise depend on the database version having been upgraded to version 10 at that point, but it won't be.

We decided to go ahead and leave this in for now as the plan is to address it later by a more robust general reindexing solution.

davecgh reviewed (changes requested):
Round 2.

I've done quite a bit of testing at this point including very large reorgs and the functionality in terms of correctness is looking great thus far.

I still need to dig into one or two more areas for a final round of review, but here is some more feedback in the mean time.

rstaudt2 commented on blockchain/compress.go:
```diff
@@ -684,85 +660,63 @@ func decodeCompressedTxOut(serialized []byte, compressionVersion uint32,
-	// txTypeShift is the number of bits to shift flags to the right to yield the
-	// correct integer value after applying the bitmask with AND.
-	txTypeShift = 2
+	// txOutFlagTxTypeBitmask describes the bitmask that yields bits 4-7 from
```
Good catch, fixed.

rstaudt2 commented on blockchain/chainio.go:
```diff
@@ -481,63 +472,69 @@ func dbMaybeStoreBlock(dbTx database.Tx, block *dcrutil.Block) error {
+
+	// ticketMinOuts is the minimal outputs for the ticket transaction that the
+	// output is contained in.  This is only stored in ticket submission outputs
+	// and is nil for all other output types.
```
Ya, I think that is a good point.  Added that here and also in the `UtxoEntry` type definition.

rstaudt2 commented on blockchain/chainio.go:
```diff
@@ -571,94 +563,67 @@ func spentTxOutSerializeSize(stxo *spentTxOut) int {
+	if offset >= len(serialized) {
+		return offset, errDeserialize("unexpected end of data after flags")
 	}
+	isCoinBase, hasExpiry, txType := decodeFlags(txOutFlags(flags))
```
No, I think that was just an artifact of reusing code from the utxo entries where the in-memory encoding is different than the serialized encoding.  Updated this as suggested.

rstaudt2 commented on blockchain/chainio.go:
```diff
@@ -850,181 +798,155 @@ func dbRemoveSpendJournalEntry(dbTx database.Tx, blockHash *chainhash.Hash) erro
-// optimized to reduce space using domain specific compression algorithms.  This
-// format is a slightly modified version of the format used in Bitcoin Core.
+// unspent output using a format that is optimized to reduce space using domain
+// specific compression algorithms.  This format is a slightly modified version
```
Removed.

rstaudt2 commented on blockchain/chainio.go:
```diff
@@ -850,181 +798,155 @@ func dbRemoveSpendJournalEntry(dbTx database.Tx, blockHash *chainhash.Hash) erro
+
+// outpointFromKey returns the outpoint that the provided serialized key
+// represents.
+func outpointFromKey(serialized []byte) (*wire.OutPoint, error) {
```
That is a good point, updated.

rstaudt2 commented on blockchain/chainio.go:
```diff
@@ -1061,133 +976,71 @@ func deserializeUtxoEntry(serialized []byte) (*UtxoEntry, error) {
+	// Copy the minimal outputs if this was a ticket submission output.
+	if isTicketSubmissionOutput(txType, txOutIndex) {
+		entry.ticketMinOuts = &ticketMinimalOutputs{
+			data: make([]byte, len(serialized[offset:])),
```
The serialized minimal outputs actually do already start with a VLQ for the number of outputs.  I was actually already using this to check the size for the stxos, and updated to do the same here for utxos now.  Thanks for pointing that out.

rstaudt2 commented on internal/mempool/mempool_test.go:
```diff
@@ -83,16 +83,20 @@ func (s *fakeChain) FetchUtxoView(tx *dcrutil.Tx, treeValid bool) (*blockchain.U
 	// do not affect the fake chain's view.
 
-	// Add an entry for the tx itself to the new view.
+	// Add entries for the outputs of the tx itself to the new view.
```
Updated here and also in `mining_harness_test.go`.

davecgh commented on blockchain/chainio_test.go:
```diff
@@ -557,227 +502,218 @@ func TestSpendJournalSerialization(t *testing.T) {
+			LockTime: 0,
+			Expiry:   0,
+		}},
+		serialized: hexToBytes("0100610511db93e1dcdb8a016b49840f8c53bc1eb68a382e9" +
```
Alright, I see what happened.  The scripts in the mock entries do not match the scripts the transactions are actually spending as they should.  Referencing the original PR this ported, I can see the `pkScript` for this one is supposed to be the uncompressed `410411db93e1dcdb8a016b49840f8c53bc1eb68a382e97b1482ecad7b148a6909a5cb2e0eaddfb84ccf9744464f82e160bfa9b8b64f9d4c03f999b8643f656b412a3ac` and then the serialized output is `01000511db93e1dcdb8a016b49840f8c53bc1eb68a382e97b1482ecad7b148a6909a5c` as expected.

The rest need to be updated accordingly.

My comment about the stake scripts being given their own `cst` types is still true.

rstaudt2 replied:
Thanks for pointing that out.  I ended up reworking this test to have a more realistic case that also covers all of the various code paths in the spend journal serialization/deserialization.  I did confirm that the regular tx is compressed and the stake tx is not (as expected, like you described).

sefbkn commented on blockchain/utxoviewpoint.go:
```diff
@@ -488,25 +509,63 @@ func (view *UtxoViewpoint) disconnectTransactions(block *dcrutil.Block, stxos []
+		hasExpiry := msgTx.Expiry != wire.NoExpiryValue
+
+		// Mark all of the spendable outputs originally created by the transaction
+		// as spent.  It is instructive to note that while the outputs aren't
```
This was one of the changes I suggested since the original seemed hard to parse. The above rewording reads much clearer.

rstaudt2 replied:
Updated to the suggested wording.

davecgh commented on blockchain/chainio_test.go:
```diff
@@ -895,293 +831,158 @@ func TestUtxoSerialization(t *testing.T) {
-		},
-		{
-			name: "outputs 0 and 2 not coinbase",
+			serialized: hexToBytes("df3982a7310132000496b538e853519c726a2c91e61ec11" +
```
I have manually verified all of these serializations in `TestUtxoSerialization` are accurate.

rstaudt2 commented on blockchain/utxoviewpoint.go:
```diff
@@ -515,72 +574,64 @@ func (view *UtxoViewpoint) disconnectTransactions(block *dcrutil.Block, stxos []
 // information.
-func (view *UtxoViewpoint) disconnectRegularTransactions(block *dcrutil.Block, stxos []spentTxOut, isTreasuryEnabled bool) error {
-	return view.disconnectTransactions(block, stxos, false, isTreasuryEnabled)
+func (view *UtxoViewpoint) disconnectRegularTransactions(db database.DB, block *dcrutil.Block, stxos []spentTxOut, isTreasuryEnabled bool) error {
```
Removed `db`.

rstaudt2 commented on blockchain/utxoviewpoint.go:
```diff
@@ -515,72 +574,64 @@ func (view *UtxoViewpoint) disconnectTransactions(block *dcrutil.Block, stxos []
 // information.
-func (view *UtxoViewpoint) disconnectStakeTransactions(block *dcrutil.Block, stxos []spentTxOut, isTreasuryEnabled bool) error {
-	return view.disconnectTransactions(block, stxos, true, isTreasuryEnabled)
+func (view *UtxoViewpoint) disconnectStakeTransactions(db database.DB, block *dcrutil.Block, stxos []spentTxOut, isTreasuryEnabled bool) error {
```
Removed `db`.

rstaudt2 commented on blockchain/utxoviewpoint.go:
```diff
@@ -454,15 +474,15 @@ func (view *UtxoViewpoint) connectTransaction(tx *dcrutil.Tx, blockHeight int64,
+// transactions in either the regular or stake tree of the block, depending on
+// the flag, and unspending all of the txos spent by those same transactions by
+// using the provided spent txo information.
+func (view *UtxoViewpoint) disconnectTransactions(db database.DB, block *dcrutil.Block, stxos []spentTxOut, stakeTree bool, isTreasuryEnabled bool) error {
```
Removed `db`.

rstaudt2 commented on blockchain/utxoviewpoint.go:
```diff
@@ -515,72 +574,64 @@ func (view *UtxoViewpoint) disconnectTransactions(block *dcrutil.Block, stxos []
-func (view *UtxoViewpoint) disconnectStakeTransactions(block *dcrutil.Block, stxos []spentTxOut, isTreasuryEnabled bool) error {
-	return view.disconnectTransactions(block, stxos, true, isTreasuryEnabled)
+func (view *UtxoViewpoint) disconnectStakeTransactions(db database.DB, block *dcrutil.Block, stxos []spentTxOut, isTreasuryEnabled bool) error {
+	return view.disconnectTransactions(db, block, stxos, true, isTreasuryEnabled)
```
Removed `db`.

rstaudt2 commented on blockchain/utxoviewpoint.go:
```diff
@@ -515,72 +574,64 @@ func (view *UtxoViewpoint) disconnectTransactions(block *dcrutil.Block, stxos []
-func (view *UtxoViewpoint) disconnectRegularTransactions(block *dcrutil.Block, stxos []spentTxOut, isTreasuryEnabled bool) error {
-	return view.disconnectTransactions(block, stxos, false, isTreasuryEnabled)
+func (view *UtxoViewpoint) disconnectRegularTransactions(db database.DB, block *dcrutil.Block, stxos []spentTxOut, isTreasuryEnabled bool) error {
+	return view.disconnectTransactions(db, block, stxos, false, isTreasuryEnabled)
```
Removed `db`.

rstaudt2 commented on blockchain/utxoviewpoint.go:
```diff
@@ -616,7 +666,7 @@ func (view *UtxoViewpoint) disconnectDisapprovedBlock(db database.DB, block *dcr
 	}
 
-	return view.disconnectRegularTransactions(block, stxos, isTreasuryEnabled)
+	return view.disconnectRegularTransactions(db, block, stxos, isTreasuryEnabled)
```
Removed `db`.

rstaudt2 commented on blockchain/utxoviewpoint.go:
```diff
@@ -713,11 +763,11 @@ func (view *UtxoViewpoint) disconnectBlock(db database.DB, block, parent *dcruti
 	// the block.  Notice that the regular tree is disconnected before the stake
 	// tree since that is the reverse of how they are connected.
-	err = view.disconnectRegularTransactions(block, stxos, isTreasuryEnabled)
+	err = view.disconnectRegularTransactions(db, block, stxos, isTreasuryEnabled)
```
Removed `db`.

rstaudt2 commented on blockchain/utxoviewpoint.go:
```diff
@@ -713,11 +763,11 @@ func (view *UtxoViewpoint) disconnectBlock(db database.DB, block, parent *dcruti
 		return err
 	}
-	err = view.disconnectStakeTransactions(block, stxos, isTreasuryEnabled)
+	err = view.disconnectStakeTransactions(db, block, stxos, isTreasuryEnabled)
```
Removed `db`.

rstaudt2 commented on blockchain/utxoviewpoint.go:
```diff
@@ -1001,12 +1047,21 @@ func (b *BlockChain) FetchUtxoView(tx *dcrutil.Tx, includeRegularTxns bool) (*Ut
 	// Create a set of needed transactions based on those referenced by the
-	// inputs of the passed transaction.  Also, add the passed transaction
+	// inputs of the passed transaction.  Also, add the outputs of the transaction
 	// itself as a way for the caller to detect duplicates that are not fully
```
Updated.

rstaudt2 commented on blockchain/utxoviewpoint.go:
```diff
@@ -204,40 +201,27 @@ func (entry *UtxoEntry) Clone() *UtxoEntry {
+
+	if entry.ticketMinOuts != nil {
+		newEntry.ticketMinOuts = &ticketMinimalOutputs{
+			data: make([]byte, len(entry.ticketMinOuts.data)),
```
I confirmed that the `data` within `ticketMinOuts` is never directly modified and updated the clone function to just copy the `ticketMinOuts` pointer.

rstaudt2 commented on blockchain/chainio_test.go:
```diff
@@ -449,88 +444,38 @@ func TestStxoDecodeErrors(t *testing.T) {
-		},
-	}
+	}{{
+		name:       "nothing serialized (no flags)",
```
I adjusted these tests, as well as the utxo deserialization error tests, to be more thorough and also added comments indicating where the failure is intended to be.

rstaudt2 commented on blockchain/chainio_test.go:
```diff
@@ -347,53 +346,46 @@ func TestStxoSerialization(t *testing.T) {
+			blockHeight: 12345,
+			blockIndex:  54321,
+			packedFlags: encodeFlags(
+				true,  // coinbase
```
Agreed, that is a bit clearer.  Updated.

rstaudt2 commented on blockchain/chainio_test.go:
```diff
@@ -449,88 +444,38 @@ func TestStxoDecodeErrors(t *testing.T) {
+		serialized: hexToBytes("010012"),
+		errType:    errDeserialize(""),
+		bytesRead:  2,
+	}}
```
Good call.  I updated these tests, as well as some of the other serialization/deserialization tests, to have more comprehensive coverage.  They now have pretty much full coverage, with the only missing cases being code paths that can't be hit (in the case of `decodeCompressedTxOut` it could be hit if called directly, but when called from the deserialize functions they already check for that case before calling it).
```
$ go test -run="TestStxoSerialization|TestStxoDecodeErrors|TestSpendJournalSerialization|TestSpendJournalErrors|TestUtxoSerialization|TestUtxoEntryDeserializeErrors" -coverprofile=cov.out
$ go tool cover -func=cov.out | grep -E "spentTxOutSerializeSize|putSpentTxOut|decodeSpentTxOut|deserializeSpendJournalEntry|serializeSpendJournalEntry|serializeUtxoEntry|deserializeUtxoEntry|encodeFlags|decodeFlags|compressedTxOutSize|putCompressedTxOut|decodeCompressedTxOut|readDeserializeSizeOfMinimalOutputs"
github.com/decred/dcrd/blockchain/v4/chainio.go:209:		readDeserializeSizeOfMinimalOutputs	100.0%
github.com/decred/dcrd/blockchain/v4/chainio.go:549:		spentTxOutSerializeSize			100.0%
github.com/decred/dcrd/blockchain/v4/chainio.go:569:		putSpentTxOut				100.0%
github.com/decred/dcrd/blockchain/v4/chainio.go:589:		decodeSpentTxOut			100.0%
github.com/decred/dcrd/blockchain/v4/chainio.go:642:		deserializeSpendJournalEntry		100.0%
github.com/decred/dcrd/blockchain/v4/chainio.go:704:		serializeSpendJournalEntry		93.8%
github.com/decred/dcrd/blockchain/v4/chainio.go:926:		serializeUtxoEntry			100.0%
github.com/decred/dcrd/blockchain/v4/chainio.go:964:		deserializeUtxoEntry			100.0%
github.com/decred/dcrd/blockchain/v4/compress.go:583:		compressedTxOutSize			100.0%
github.com/decred/dcrd/blockchain/v4/compress.go:601:		putCompressedTxOut			100.0%
github.com/decred/dcrd/blockchain/v4/compress.go:621:		decodeCompressedTxOut			86.4%
github.com/decred/dcrd/blockchain/v4/compress.go:698:		encodeFlags				100.0%
github.com/decred/dcrd/blockchain/v4/compress.go:714:		decodeFlags				100.0%
```

rstaudt2 commented on blockchain/chainio_test.go:
```diff
@@ -1200,97 +1001,26 @@ func TestUtxoSerialization(t *testing.T) {
+		// Don't try to deserialize if the test entry was spent since it will have a
+		// nil serialization.
+		if test.entry.IsSpent() {
+			if gotBytes != nil {
```
Agreed, removed.

davecgh commented on blockchain/upgrade.go:
```diff
@@ -1789,6 +1789,1162 @@ func upgradeToVersion8(ctx context.Context, db database.DB, dbInfo *databaseInfo
+						"data after script size (got %v, need %v)",
+						len(oldSerialized[offset:]), scriptSize))
+				}
+				compressedScript := make([]byte, scriptSize)
```
It probably won't make much difference in the overall time and/or memory given most entries are pretty small, but the script really doesn't need to be duplicated here.  It's only duplicated in the normal path because the utxo is later returned and thus no reference into the data can be kept so the serialized data can be GCd, but that doesn't apply here because it's only live for the duration of a single record.

```suggestion
				compressedScript := oldSerialized[offset:offset+scriptSize]
```

rstaudt2 replied:
Good call, updated.

davecgh commented on blockchain/upgrade.go:
```diff
@@ -1789,6 +1789,1162 @@ func upgradeToVersion8(ctx context.Context, db database.DB, dbInfo *databaseInfo
+			// Copy the stake extra data if this was a ticket.
+			var ticketMinOuts []byte
+			if txType == stake.TxTypeSStx {
+				ticketMinOuts = make([]byte, len(oldSerialized[offset:]))
```
Same here.

rstaudt2 replied:
Updated here and a few other places within the migration that did not require copying as well.

davecgh commented on blockchain/upgrade.go:
```diff
@@ -1789,6 +1789,1162 @@ func upgradeToVersion8(ctx context.Context, db database.DB, dbInfo *databaseInfo
+			//   Field                Type             Size
+			//   hash                 chainhash.Hash   chainhash.HashSize
+			//   output index         VLQ              variable
+			//   tree                 VLQ              variable
```
Do you think it would be worth making the order `<hash><tree><output index>` instead so it's possible to iterate over a specific tree by creating a cursor with the prefix `<hash><tree>` and also the natural iteration order would then be sorted by tree, then block index, too?

It might come in handy for the cache and it seems like it would more logically match how they're actually structured within a block and thus likely to be looked up, perhaps allowing for faster sequential seeking versus the current layout that will require randomly jumping back and forth as it first looks up outputs from the regular tree and then the stake tree.

e.g. Currently:

```
<hash a><block index 0><tree 0>  <-- Loaded 1st
<hash a><block index 0><tree 1>  <-- Loaded 4th
<hash a><block index 1><tree 0>  <-- Loaded 2nd
<hash a><block index 1><tree 1>  <-- Loaded 5th
<hash a><block index 2><tree 0>  <-- Loaded 3rd
<hash a><block index 2><tree 1>  <-- Loaded 6th
```

Versus:
```
<hash a><tree 0><block index 0>  <-- Loaded 1st
<hash a><tree 0><block index 1>  <-- Loaded 2nd
<hash a><tree 0><block index 2>  <-- Loaded 3rd
<hash a><tree 1><block index 0>  <-- Loaded 4th
<hash a><tree 1><block index 1>  <-- Loaded 5th
<hash a><tree 1><block index 2>  <-- Loaded 6th
```

I realize this would apply elsewhere to the new v3 format and not just here.

This is definitely not something that is necessary to land this PR at all, just a thought.

rstaudt2 replied:
I thought about this a bit and went back and forth on it (also, it should be `output index` in your comment rather than `block index`).  On the one hand, I don't think it really matters either way, since:

- In practice, having the same hash for both trees would be extremely rare, so jumping back and forth to check both trees would never really happen.
- We want to avoid seeking in general, and we've removed all cases that required this up to this point and should be able to lookup by a proper outpoint key.

That being said, I decided to switch it to `<hash><tree><index>` for the following reasons:

- In the case where we know the `hash` but don't know the `tree` and/or `index`, I think it's preferred to have the `tree` come first.
    - In the case you have the `tree` but don't have the `index`:
        - You can easily scan `<hash><tree>...` to find the first index or all indexes, depending on what you are looking for.
    - In the case that you have the `index`, but don't have the `tree`:
        - You would need to check both trees, but since there are only 2 trees, this is still efficient.
- If we have the `index` come first, the main downside is the case where you know the `tree` but do not know the `index`:
    - Scanning would be a bit less straightforward in this case, as you'd have to scan `<hash><each output>` to find the first output, but then check the `tree` when an entry is found, and continue scanning if the `tree` doesn't match (and even though it would be extremely unlikely that the tree wouldn't match, we'd still have to check and handle that case).
- I do think that the structure makes more sense the `hash` and `tree` are used to identify the tx, whereas the index is used to identify the output, e.g.:
    ```
    |-tx unique identifier-| |---- tx output ----|
          <hash><tree>           <output index>
    ```

Since it's a bit cleaner and a bit more future proof (and was a small update), I went ahead and made that change.

davecgh replied:
Great.  Indeed I originally said output index and then put block index in the following examples.  Oops.

I agree that we typically shouldn't generally be scanning/seeking anymore since we should be doing all lookups by the full outpoint.  However, I also know that sometimes things come up that require limited scanning and it's much harder to change keys in the database later.

As you note, the number of trees is significantly less than the output indexes (2 versus potentially 2^32-1) , so it significantly limits the search space in the event you don't have the full outpoint for whatever reason.

Come to think of it, I think we already have one such case when trying to determine if inventory is already known by falling back to the utxoset which ends up looking into both trees for a limited number of outputs.  It's not something to worry about for this PR, but that could probably be optimized now due to this change to do a cursor lookup of `<hash><tree>` for the two trees instead to determine if there are _any_ outputs.

davecgh reviewed (commented):
The only outstanding issue as far as I can tell is the migration code issue surrounding side chain blocks.  Once that is in, I'll give it another round of review and testing and assuming that goes well, this will be ready.

rstaudt2 commented on blockchain/chainio.go:
```diff
@@ -850,181 +801,158 @@ func dbRemoveSpendJournalEntry(dbTx database.Tx, blockHash *chainhash.Hash) erro
+//   flags                VLQ      variable
+//   compressed txout
 //     compressed amount   VLQ      variable
 //     compressed version  VLQ      variable
```
Good call, updated.

rstaudt2 commented:
> The only outstanding issue as far as I can tell is the migration code issue surrounding side chain blocks. Once that is in, I'll give it another round of review and testing and assuming that goes well, this will be ready.

This should be resolved now, did further testing around side chain blocks for the migration as well as invalidating/reconsidering the main chain block at the same height as a side chain block.

davecgh reviewed (approved):
All updates reviewed and test.

Migration stats:

```
2021-01-10 15:26:08.780 [INF] CHAN: Done migrating utxoset.  Total entries: 660256 in 14.265s
2021-01-10 15:26:08.780 [INF] CHAN: Removing old utxoset entries...
2021-01-10 15:26:11.109 [INF] CHAN: Deleted 660256 keys (660256 total) from old utxoset
2021-01-10 15:26:11.109 [INF] CHAN: Done removing old utxoset entries in 2.328s
...
2021-01-10 15:41:54.722 [INF] CHAN: Done migrating spend journal.  Total entries: 496670 in 15m37.916s
2021-01-10 15:41:54.722 [INF] CHAN: Removing temp data...
...
2021-01-10 15:42:39.690 [INF] CHAN: Done removing temp data in 44.968s
...
2021-01-10 15:42:43.155 [INF] CHAN: Done removing old spend journal in 3.464s
2021-01-10 15:42:43.155 [INF] CHAN: Done upgrading database in 16m48.644s.
2021-01-10 15:42:43.158 [INF] CHAN: Blockchain database version info: chain: 9, compression: 1, block index: 3
...
```

Memory usage peaked around 800MB.

matheusd commented on blockchain/upgrade.go:
```diff
@@ -1789,6 +1789,1184 @@ func upgradeToVersion8(ctx context.Context, db database.DB, dbInfo *databaseInfo
+						continue
+					}
+
+					// If the temp bucked didn't have the tx info, check the utxo set for
```
bucket

rstaudt2 replied:
Fixed.

matheusd reviewed (approved):
Looks good other than a typo.
